### DISCUSSION TITLE
Judgements are handled for update time rather than input time

### DISCUSSION THREAD

**Discussion by Firmatorenio** [ppy/osu#26457] [Q&A]
I was fiddling with refresh rate multipliers and just happened to do so on a map with a bpm that was also a multiplier of my refresh rate, and happened to notice a similar pattern on the hit error bar that is often demonstrated when testing a keyboard for bad polling rates.

![image](https://github.com/ppy/osu/assets/12499153/cc3cc94c-78cb-4201-9ba0-80893c7cb8ed)
This is the picture I saw playing a 240bpm map at 2x refresh rate, with update thread at 240hz, notice how evenly the judgements are spaced.

![image](https://github.com/ppy/osu/assets/12499153/06a9e8d0-b033-4b99-adec-062c99a99557)
This is what I see if I bump the refresh rate multiplier to 8x, and how the hit error bar is supposed to look.

A similar story can be visible on the timing distribution graph, here's what I see with 2x refresh rate:
![image](https://github.com/ppy/osu/assets/12499153/b14abc02-a102-4dae-a501-db0a863f3427)

8x yields a much more normal-looking one:
![image](https://github.com/ppy/osu/assets/12499153/af1b4b8b-1244-46c7-8292-d49f3c2c8fbc)

I have tested this with several keyboards with different polling/report rates, using different ones did not seem to affect the result at all. My specs are i5-11400f, Arc A750, 16GB RAM, the game is installed on an NVME PCIE4.0 SSD while the database is on a SATA3 SSD. Unfortunately I do not have any monitors with refresh rates higher than 60hz to test them.

[Here](https://youtu.be/j6vU8YIiIms)'s a full video and steps reproducing this phenomenon:
- Ensure you're on multithreaded mode, in exclusive fullscreen and on 2x refresh rate, make sure your CPU can keep up with the update thread (has a stable 240hz for 60hz monitors);
- Find a map with a straight beat and a bpm that is a multiple of your refresh rate (120 or 240 bpm for 60hz, alternatively you can play a 160bpm map with a swing beat or a 160bpm map with a straight beat with DT/NC at 1.5x, I tested both those cases too);
- (optional, makes the pattern more visual) Set your OD to 10 (or even 11);
- Play through the map normally, observe the hit error meter during the play as well as the hit distribution graph at the extended results screen having evenly spaced patterns;
- Set the refresh rate multiplier to 8x, once again make sure your CPU can keep up;
- Play the same map again, observe the patterns not emerge in either place.

Digging into the code, it would seem that the `DrawableHitObject` code uses `Current` time when judging hits rather than input time, and I would assume that code has to be running on an update thread? If it _is_ run on the update thread, wouldn't it want to use whatever time the input was handled at rather than whatever time it is when it's processed?
[`Result.RawTime = Time.Current;`](https://github.com/ppy/osu/blob/f376bb5ec7ba6ed8cf609dac67cd96a7123c03ca/osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs#L706C43-L706C43)

I don't know if this is an issue with just my machine (which is the reason this is a discussion and not an issue), and the 4ms latency might not be all that significant, but given one of lazer's selling points is the 1000hz input thread, saying there's no input lag even at low framerates may be a little disingenuous.

**Comment by peppy** [comment]
I'm 100% sure there's an issue open tracking this. It's a framework side issue which requires support for timestamped input. Will come later.

FWIW the input thread means nothing right now if you are using raw input, as that is event driven.

**Reply by Firmatorenio** [reply]
I couldn't find anything on osu side, but I do see ppy/osu-framework#4808 with a broken twitter link now. I personally think this should be way higher priority with the whole ranked play thing right around the corner.

**Reply by peppy** [reply]
That's the one!

Thanks for your advice on prioritisation! We have other things to handle first though. You can sooner expect to see the update frame time go down, as we continue to optimise the game.
